File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/auto/PipelineRed.java
package org.firstinspires.ftc.teamcode.auto;

import org.opencv.core.*;
import org.opencv.core.Mat;
import org.opencv.core.MatOfPoint;
import org.opencv.core.MatOfPoint2f;
import org.opencv.core.Rect;
import org.opencv.core.Scalar;
import org.opencv.imgproc.Imgproc;
import org.opencv.imgproc.Moments;
import org.openftc.easyopencv.OpenCvPipeline;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;

// have JunctionPipeline return an x coordinate and a y coordinate in a list, maybe?

public class PipelineRed extends OpenCvPipeline {

    public Scalar darkestJunctions = new Scalar(0, 167, 85);
    public Scalar lightestJunctions = new Scalar(255, 255, 128);
    public double smallestArea = 4000;
    List<Mat> channels = new ArrayList<>();
    Mat rawYCrCb = new Mat();
    Mat blurred = new Mat();
    Mat blueMat = new Mat();
    Mat thresholded = new Mat();
    int junctionNumAttr = 0;
    Point junctionPointAttr = new Point();
    double junctionDistanceAttr = 0;
    double propAreaAttr;

    public Mat processFrame(Mat input) {

        // crop out parts we're not concerned about

        // Convert image to HSV
        Imgproc.cvtColor(input, rawYCrCb, Imgproc.COLOR_RGB2YCrCb);

        // Blur image to lessen noise
        Imgproc.GaussianBlur(rawYCrCb, blurred, new Size(15, 15), 0); // increase blur?

        Core.inRange(blurred, darkestJunctions, lightestJunctions, thresholded);

        /*
        Core.split(input, channels);
        blueMat = channels.get(0);

        // Threshold image, turning it into binary (only black and white). Now openCV knows what to get the contour, or shape, of.
        Core.inRange(blueMat, darkestJunctions, lightestJunctions, thresholded);
        */


        // Find contours
        List<MatOfPoint> contours = new ArrayList<>();
        Imgproc.findContours(thresholded, contours, new Mat(), Imgproc.RETR_EXTERNAL, Imgproc.CHAIN_APPROX_SIMPLE);

        // Get distance and centroid of biggest junction
        List<MatOfPoint> bigContours = new ArrayList<>();
        if (!contours.isEmpty()) {
            MatOfPoint biggestContour = contours.get(0);

            for (MatOfPoint curContour : contours) {
                if (Imgproc.contourArea(curContour) > Imgproc.contourArea(biggestContour)) {
                    biggestContour = curContour;
                }
            }
            bigContours.add(biggestContour);

            // Find centroid
            Moments moments = Imgproc.moments(biggestContour);
            Point junctionPoint = new Point(moments.get_m10() / moments.get_m00(), moments.get_m01() / moments.get_m00());

            //Assign attributes
            junctionNumAttr = contours.size();
            propAreaAttr = Imgproc.contourArea(biggestContour);
            junctionDistanceAttr = 240000/Imgproc.contourArea(biggestContour);
            junctionPointAttr = junctionPoint;
        }

        Imgproc.drawContours(input, contours, -1, new Scalar(0,255,0), 3);
        Imgproc.drawContours(input, bigContours, -1, new Scalar(255,0,0), 3);

        return input;
    }

    public Point getJunctionPoint() {
        return junctionPointAttr;
    }

    public int getJunctionNum() {return junctionNumAttr;}
    public double getPropAreaAttr() {return propAreaAttr;}
    public double getJunctionDistance() {
        return junctionDistanceAttr; // this is in inches
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/vision/PipelineRed.kt
package org.firstinspires.ftc.teamcode.vision

import org.opencv.core.Core
import org.opencv.core.Mat
import org.opencv.core.Rect
import org.opencv.core.Scalar
import org.opencv.imgproc.Imgproc
import org.openftc.easyopencv.OpenCvPipeline

class PipelineRed : OpenCvPipeline(){

    var workingMatrix = Mat()

    var cupState = CupStates.RIGHT
    enum class CupStates {
        LEFT, CENTER, RIGHT,
    }

    var MIN_R = 60

    var LeftTotal : Double = 0.0
    var CenterTotal : Double = 0.0
    var RightTotal : Double = 0.0


    override fun processFrame (input : Mat) : Mat {
        input.copyTo(workingMatrix)

        if (workingMatrix.empty()) {
            return input
        }

        val matLeft = workingMatrix.submat(239, 240, 319, 320)
        val matCenter = workingMatrix.submat(180, 230, 180, 240)
        val matRight = workingMatrix.submat(180, 230, 0, 60)

        LeftTotal =  Core.mean(matLeft).`val`[2]
        CenterTotal = Core.mean(matCenter).`val`[2]
        RightTotal = Core.mean(matRight).`val`[2]

        Imgproc.cvtColor(workingMatrix, workingMatrix, Imgproc.COLOR_RGB2YCrCb)

        Imgproc.rectangle(workingMatrix, Rect(319, 239, 1, 1), Scalar(0.0, 255.0, 0.0))
        Imgproc.rectangle(workingMatrix, Rect(180, 180, 60, 50), Scalar(255.0, 0.0, 0.0))
        Imgproc.rectangle(workingMatrix, Rect(0, 180, 60, 50), Scalar(255.0, 0.0, 0.0))


        return workingMatrix
    }
}